llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1976,8 +1976,11 @@ struct VPCSEDenseMapInfo : public DenseMapInfo<VPSingleDefRecipe *> {
   static bool canHandle(const VPSingleDefRecipe *Def) {
     // We can extend the list of handled recipes in the future,
     // provided we account for the data embedded in them while checking for
-    // equality or hashing.
-    auto C = getOpcodeOrIntrinsicID(Def);
+    // equality or hashing. We assign VPVectorEndPointerRecipe the GEP opcode,
+    // as it is essentially a GEP with different semantics.
+    auto C = isa<VPVectorPointerRecipe>(Def)
```

Comment:
Where are we checking for the indexed type? On VPVectorPointerRecipe it's stored in the class, and I think for VPWidenGEPRecipe it gets it from the underlying instruction?

Reply:
Thanks! I didn't realize that this was the source element type of the GEP: I thought for some reason that all GEPs were canonicalized to have the i8 source element type, due to recent changes.

Reply:
This is very hard to test: if you look at what's being eliminated, it's mostly unrolled GEPs.

Reply:
> This is very hard to test: if you look at what's being eliminated, it's mostly unrolled GEPs.

Is it? I think something like https://llvm.godbolt.org/z/3zvK7TTos should do

Reply:
Damn, I totally forgot replicates, and was trying to craft a test with WidenGEP/VectorEndPointer: sorry for the shoddiness; will fix in the morning.